src/ray/object_manager/object_manager.cc

```diff
@@ -196,19 +205,34 @@ ray::Status ObjectManager::PullSendRequest(const ObjectID &object_id,
   return ray::Status::OK();
 }
 
-ray::Status ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id,
-                                int retry) {
-  if (local_objects_.count(object_id) == 0) {
-    if (retry < 0) {
-      retry = config_.max_push_retries;
-    } else if (retry == 0) {
-      RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
-                     << " after retrying " << config_.max_push_retries << " times.";
-      return ray::Status::OK();
+void ObjectManager::HandlePushTaskTimeout(const ObjectID &object_id,
+                                          const ClientID &client_id) {
+  RAY_LOG(ERROR) << "Invalid Push request ObjectID: " << object_id
+                 << " after waiting for " << config_.push_timeout_ms << " ms.";
+  auto iter = unfulfilled_push_tasks_.find(object_id);
+  if (iter != unfulfilled_push_tasks_.end()) {
+    iter->second.erase(client_id);
+    if (iter->second.size() == 0) {
+      unfulfilled_push_tasks_.erase(iter);
     }
-    main_service_->post([this, object_id, client_id, retry]() {
-      RAY_CHECK_OK(Push(object_id, client_id, retry - 1));
-    });
+  }
+}
+
+ray::Status ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id) {
+  if (local_objects_.count(object_id) == 0) {
```

Comment:
do not set up timer again if unfulfilled_push_tasks_[object_id][client_id] already exists.

Reply:
Nice Catch!